wikimedia/wikipedia-ios
# Fix remote notification API errors for variants

**Fixes Phabricator ticket:** https://phabricator.wikimedia.org/T277608

### Notes
This fixes the API errors around remote notifications and variants. Let's lean on QA to regression test any user-facing side of this.

### Test Steps
1. Go to `RemoteNotificationsOperationsController.swift`. Comment out this chunk, this will allow you to see the notifications endpoint call more often:
```
//        guard deadlineController?.isBeforeDeadline ?? false else {
//            completeEarly()
//            return
//        }
```
2. Run app, add app languages with variants in Settings. Inspect the notifications call (...`api.php?action=query&meta=notifications`...), confirm the query item for `notwikis` uses the Wiki language only for it's value, not the entire variant code: `notwikis=srwiki|enwiki|zhwiki|wikidatawiki`. In `main` it sends `notwikis=enwiki|zh_hanswiki|sr_ecwiki|wikidatawiki` and in the response we get the warning:

```
{
  "batchcomplete": true,
  "warnings": {
    "notifications": {
      "warnings": "Unrecognized values for parameter \"notwikis\": zh_hanswiki, sr_ecwiki"
    }
  },
```

3. On `main`, I would occasionally see a token error when it tried to mark a notification as read. The call looks like this:

POST https://www.mediawiki.org/w/api.php?action=echomarkread&format=json&list={{notificationID}}&wikis=enwiki
Body: token=%3CWMFToken:%200x600002e65500%3E

Response:

```
{
  "error": {
    "code": "badtoken",
    "info": "Invalid CSRF token.",
    "*": "See https://www.mediawiki.org/w/api.php for API usage. Subscribe to the mediawiki-api-announce mailing list at &lt;https://lists.wikimedia.org/mailman/listinfo/mediawiki-api-announce&gt; for notice of API deprecations and breaking changes."
  },
  "servedby": "mw1285"
}
```

Fixing that one line in `RemoteNotificationsAPIController.swift` to `token.value` fixed this API error. It's a little difficult to test this. I would:

1. Make a change to an article from the app on EN Wikipedia. 
2. On desktop, log into a different account, go to that article, go to history and click "undo" next to the revision you just made from the device.
3. Wait a while on device watching the calls, eventually after a minute or two you will see a call that tries to mark that edit revert notification as read.
